AdminServer/appscale/admin/__init__.py

```diff
@@ -310,6 +317,28 @@ def begin_deploy(self, project_id, source_path):
       message = 'Error while updating application: {}'.format(error)
       raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
 
+  def identify_as_hoster(self, project_id, source_location):
+    """ Marks this machine as having a version's source code.
+
+    Args:
+      project_id: A string specifying a project ID.
+      source_path: A string specifying the location of the version's source.
+    """
+    private_ip = appscale_info.get_private_ip()
+    hoster_node = '/apps/{}/{}'.format(project_id, private_ip)
+
+    try:
+      self.zk_client.create(hoster_node, str(source_location), ephemeral=True,
+                            makepath=True)
+    except NodeExistsError:
+      self.zk_client.set(hoster_node, str(source_location))
+
+    # Remove other hosters that have old code.
+    hosters = self.zk_client.get_children('/apps/{}'.format(project_id))
+    old_hosters = [hoster for hoster in hosters if hoster != private_ip]
+    for hoster in old_hosters:
+      self.zk_client.delete('/apps/{}/{}'.format(project_id, hoster))
```

Comment:
@jeanleonov Unfortunately, Kazoo's recursive delete implementation is not a single call. It makes a delete call for every child.